src/storage/secondary/block/rle_block_builder.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use std::borrow::Borrow;
+
+use bytes::BufMut;
+use risinglight_proto::rowset::BlockStatistics;
+
+use super::BlockBuilder;
+use crate::array::Array;
+
+/// Encodes fixed-width data into a block with run-length encoding. The layout is
+/// rle counts and data from other block builder
+/// ```plain
+/// | rle_counts_num (u32) | rle_count (u16) | rle_count | data | data | (may be bit) |
+/// ```
+pub struct RleBlockBuilder<A, B>
+where
+    A: Array,
+    B: BlockBuilder<A>,
+{
+    block_builder: B,
+    rle_counts: Vec<u16>,
+    previous_value: Option<<A::Item as ToOwned>::Owned>,
+    target_size: usize,
+}
+
+impl<A, B> RleBlockBuilder<A, B>
+where
+    A: Array,
+    B: BlockBuilder<A>,
+{
+    pub fn new(block_builder: B, target_size: usize) -> Self {
+        Self {
+            block_builder,
+            rle_counts: Vec::new(),
+            previous_value: None,
+            target_size,
+        }
+    }
+
+    fn append_inner(&mut self, item: Option<&A::Item>) {
+        self.previous_value = item.map(|x| x.to_owned());
+        self.block_builder.append(item);
+        self.rle_counts.push(1);
```

Comment:
One optimization is to store the current `rle_count` as a u16 in the struct (instead of as last element in Vec), and push the previous count to the Vec when we append the next datum. Therefore, we can avoid bound check every time we access the last item.